spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/SparkMLPredictor.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.mllib.linalg.Vector;
+import org.apache.spark.mllib.linalg.Vectors;
+import org.apache.twill.filesystem.Location;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Common class for Spark Predictors/Classifiers. Contains common Spark ML Classifiers properties, configuartion,
+ * validation and feature extraction methods to be used in predictors/classifiers.
+ */
+public abstract class SparkMLPredictor extends SparkCompute<StructuredRecord, StructuredRecord> {
+
+  private MLPredictorConfig config;
+  private Schema outputSchema = null;
+
+  /**
+   * Config class for Predictors/Classifiers. Contains common config properties to be used in predictors/classifiers.
+   */
+  protected class MLPredictorConfig extends PluginConfig {
+
+    @Description("The name of the FileSet to load the model from.")
+    protected String fileSetName;
+
+    @Description("Path of the FileSet to load the model from.")
+    protected String path;
+
+    @Nullable
+    @Description("A comma-separated sequence of fields that needs to be used for classification/prediction.")
+    protected String featureFieldsToInclude;
+
+    @Nullable
+    @Description("A comma-separated sequence of fields that needs to be excluded from being used " +
+      "in classification/prediction.")
+    protected String featureFieldsToExclude;
+
+    @Description("The field on which to set the prediction. It will be of type double.")
+    protected String predictionField;
+
+    protected MLPredictorConfig(String fileSetName, String path, @Nullable String featureFieldsToInclude,
+                           @Nullable String featureFieldsToExclude, String predictionField) {
+      this.fileSetName = fileSetName;
+      this.path = path;
+      this.featureFieldsToInclude = featureFieldsToInclude;
+      this.featureFieldsToExclude = featureFieldsToExclude;
+      this.predictionField = predictionField;
+    }
+
+    protected void validate(Schema inputSchema) {
+      SparkUtils.validateConfigParameters(inputSchema, featureFieldsToInclude, featureFieldsToExclude, predictionField,
+                                          null);
+    }
+  }
+
+  public SparkMLPredictor(MLPredictorConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Schema inputSchema = stageConfigurer.getInputSchema();
+    Preconditions.checkArgument(inputSchema != null, "Input Schema must be a known constant.");
+    config.validate(inputSchema);
+    stageConfigurer.setOutputSchema(SparkUtils.getOutputSchema(inputSchema, config.predictionField));
+  }
+
+  public String getModelPath(SparkExecutionPluginContext context) throws IOException {
```

Comment:
instead of supplying methods that are meant to be used by the children, it is better for this class to define the basic code structure and leave as little to the child as possible. 

this means it should override the initialize and transform methods from SparkCompute and define abstract methods to fill in the gaps. Something like:

```

  public abstract void initialize(SparkExecutionPluginContext context, String modelPath);

  public abstract double predict(Vector vector);

  @Override
  public void initialize(SparkExecutionPluginContext context) throws Exception {
    super.initialize(context);
    FileSet fileSet = context.getDataset(config.fileSetName);
    Location modelLocation = fileSet.getBaseLocation().append(config.path);
    if (!modelLocation.exists()) {
      throw new IllegalArgumentException(String.format("Failed to find model to use for Regression. Location does "
                                                         "not exist: %s.", modelLocation));
    }
    initialize(context, modelLocation.toURI().getPath());
  }

    @Override
  public JavaRDD<StructuredRecord> transform(SparkExecutionPluginContext context,
                                             JavaRDD<StructuredRecord> input) throws Exception {
    if (input == null || input.isEmpty()) {
      return context.getSparkContext().emptyRDD();
    }
    Schema schema = input.first().getSchema();
    outputSchema = (outputSchema != null) ? outputSchema : SparkUtils.getOutputSchema(schema, config.predictionField);
    final Map<String, Integer> fields = SparkUtils.getFeatureList(schema, config.featuresToInclude,
                                                                  config.featuresToExclude,
                                                                  config.predictionField);
    JavaRDD<StructuredRecord> output = input.map(new Function<StructuredRecord, StructuredRecord>() {
      @Override
      public StructuredRecord call(StructuredRecord record) throws Exception {
        Vector vector = getVector(record, )
        double prediction = loadedModel.predict(vector, fields);

        return SparkUtils.cloneRecord(record, outputSchema, config.predictionField)
          .set(config.predictionField, prediction)
          .build();
      }
    });
    return output;
  }
```

This way, the child is basically a constructor and two very short methods.